presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -397,36 +399,32 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
             }
             NormalizedSimpleComparison normalized = optionalNormalized.get();
 
-            Symbol symbol = Symbol.from(normalized.getNameReference());
-            Type fieldType = checkedTypeLookup(symbol);
-            NullableValue value = normalized.getValue();
-
-            Optional<NullableValue> coercedValue = coerce(value, fieldType);
-            if (coercedValue.isPresent()) {
-                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
+            Expression symbolExpression = normalized.getSymbolExpression();
+            if (symbolExpression instanceof SymbolReference) {
+                Symbol symbol = Symbol.from(symbolExpression);
+                NullableValue value = normalized.getValue();
+                Type type = value.getType(); // common type for symbol and value
+                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, type, value.getValue(), complement);
             }
+            else if (symbolExpression instanceof Cast) {
+                Cast castExpression = (Cast) symbolExpression;
```

Comment:
Cast is being un-wrapped only once. Aren't we suppose to unwrap cast until we actually reach a symbol reference? If we assume that double coercion will be simplified by the optimizer then shouldn't we add explicit check here? To make sure that after single unwrap we actually get the symbol reference? Also if we assume so, the `checkSymbolExpressionValid` must be changed. So it won't traverse deeper that 1 level deep?